### PR TITLE
[WIP] Dev zdcrec. New format for RecEvent

### PR DESCRIPTION
**Pull request by cortesep**
First version with updated format for reconstructed events

### CODE REVIEW

File: DataFormats/Detectors/ZDC/include/DataFormatsZDC/RecEvent.h
Comment (shahor02) [inline DataFormats/Detectors/ZDC/include/DataFormatsZDC/RecEvent.h]: RecEvent is not messegeable, since it contains dynamic structures as data members.

File: DataFormats/Detectors/ZDC/include/DataFormatsZDC/RecEvent.h
Comment (shahor02) [inline DataFormats/Detectors/ZDC/include/DataFormatsZDC/RecEvent.h:73]: Since it does not need to be persistent but is just a wrapper, you don't need dictionary (and adding it to LinkDef)

File: DataFormats/Detectors/ZDC/include/DataFormatsZDC/RecEvent.h
Comment (shahor02) [inline DataFormats/Detectors/ZDC/include/DataFormatsZDC/RecEvent.h:33]: In general, the concept of `mRecBC` referring to ranges in the `mEnergy`  etc. is orthogonal to having a RecEvent as a single object. Your `RecEvent` is in fact the result of the reconstruction for the full TF, and you should not store / send it as a single object. If you want, you can keep it as a wrapper for its vectors in the reconstruction code, to make it more readable, but the reco. spec. should send these individual vectors to profit from shared memory transport.

File: DataFormats/Detectors/ZDC/include/DataFormatsZDC/RecEventAux.h
Comment (shahor02) [inline DataFormats/Detectors/ZDC/include/DataFormatsZDC/RecEventAux.h]: this is also not messageable. And, as far as I understand, this is just a debug output to be stored in the root tree, right?

File: Detectors/ZDC/reconstruction/include/ZDCReconstruction/RecoParamZDC.h
Comment (shahor02) [inline Detectors/ZDC/reconstruction/include/ZDCReconstruction/RecoParamZDC.h]: the `struct` is public by default

File: Detectors/ZDC/reconstruction/src/DigiReco.cxx
Comment (shahor02) [inline Detectors/ZDC/reconstruction/src/DigiReco.cxx]: You better use `unique_ptr`, see e.g.
https://github.com/AliceO2Group/AliceO2/blob/7a6c7f384bddc76e787b6efa32c5a0842c661cf1/Detectors/Vertexing/include/DetectorsVertexing/PVertexer.h#L155-L156
https://github.com/AliceO2Group/AliceO2/blob/7a6c7f384bddc76e787b6efa32c5a0842c661cf1/Detectors/Vertexing/src/PVertexer.cxx#L615-L616
https://github.com/AliceO2Group/AliceO2/blob/7a6c7f384bddc76e787b6efa32c5a0842c661cf1/Detectors/Vertexing/src/PVertexer.cxx#L638-L644

File: Detectors/ZDC/reconstruction/src/DigiReco.cxx
Comment (shahor02) [inline Detectors/ZDC/reconstruction/src/DigiReco.cxx]: I assume this method is called from the DigitRecoSpec, passing the spans extracted from the inputs?
Then here also one should use e.g. `gsl::span<const o2::zdc::OrbitData> orbitData` etc.

File: Detectors/ZDC/workflow/src/DigitRecoSpec.cxx
Comment (shahor02) [inline Detectors/ZDC/workflow/src/DigitRecoSpec.cxx]: Here, after doing the reconstruction, you should extract the vectors which are currently the part of the RecEvent and snapshot them one by one, e.g.
````
pc.outputs().snapshot(Output{"ZDC", "BCREC", 0, Lifetime::Timeframe}, recEvent.mRecBC);
pc.outputs().snapshot(Output{"ZDC", "ENERGY", 0, Lifetime::Timeframe}, recEvent.mEnergy);
pc.outputs().snapshot(Output{"ZDC", "TDCDATA", 0, Lifetime::Timeframe}, recEvent.mTDCData);
````


And these outputs

File: Detectors/ZDC/workflow/src/DigitRecoSpec.cxx
Comment (shahor02) [inline Detectors/ZDC/workflow/src/DigitRecoSpec.cxx:197]: The outputs I mentioned above should be declared here, e.g.
````
std::vector<OutputSpec> outputs;
outputs.emplace_back("ZDC", "BCREC", 0, Lifetime::Timeframe);
outputs.emplace_back("ZDC", "ENERGY", 0, Lifetime::Timeframe);
outputs.emplace_back("ZDC", "TDCDATA", 0, Lifetime::Timeframe);
````

File: Detectors/ZDC/workflow/src/DigitRecoSpec.cxx
Comment (shahor02) [inline Detectors/ZDC/workflow/src/DigitRecoSpec.cxx]: and used here

File: Detectors/ZDC/workflow/src/ZDCRecoWriterDPLSpec.cxx
Comment (shahor02) [inline Detectors/ZDC/workflow/src/ZDCRecoWriterDPLSpec.cxx]: here you should mention all inputs the writer should store, not only the RECBC

**Review by shahor02 — CHANGES_REQUESTED** [review]

File: DataFormats/Detectors/ZDC/include/DataFormatsZDC/RecEventAux.h
Comment (cortesep) [reply DataFormats/Detectors/ZDC/include/DataFormatsZDC/RecEventAux.h]: Removed

File: DataFormats/Detectors/ZDC/include/DataFormatsZDC/RecEvent.h
Comment (cortesep) [reply DataFormats/Detectors/ZDC/include/DataFormatsZDC/RecEvent.h]: Removed

File: DataFormats/Detectors/ZDC/include/DataFormatsZDC/RecEvent.h
Comment (cortesep) [reply DataFormats/Detectors/ZDC/include/DataFormatsZDC/RecEvent.h:73]: Leaving it for the moment. Introduced a note to remove it when going into production

File: Detectors/ZDC/reconstruction/src/DigiReco.cxx
Comment (cortesep) [reply Detectors/ZDC/reconstruction/src/DigiReco.cxx]: Changed method to use span instead of vectors

**Comment by shahor02** [comment]
@cortesep There are conflicts to resolve before the CI tests can be started

File: Detectors/ZDC/reconstruction/src/DigiReco.cxx
Comment (shahor02) [inline Detectors/ZDC/reconstruction/src/DigiReco.cxx:379]: can you use LOG instead of printf? 
Also, putting `if  (mVerbosity..)` in a tight loops over the channels will cost some performance. Since this is a debug mode for local tests only, why don't you put these outputs in e.g.
#ifdef _ZDC_DEBUG_
#endif
guards, with the debug disabled by default?

Comment (cortesep) [reply Detectors/ZDC/reconstruction/src/DigiReco.cxx:379]: I didn't know I was allowed to use ifdefs. I will follow your suggestion

Comment (cortesep) [reply Detectors/ZDC/reconstruction/src/DigiReco.cxx:379]: I prefer to use printf for certain debug output. I protected them with #ifdef O2_ZDC_DEBUG

Comment (shahor02) [reply Detectors/ZDC/reconstruction/src/DigiReco.cxx:379]: once you use it only in your local tests, this is up to you. Just note that you can have with `LOGF(INFO, "printf formats",args...)` and `LOGP(INFO, "fmt formats", args...)` the printf and fmt formated outputs resp.

**Comment by shahor02** [comment]
@cortesep 

There were lot of problems in your layout:
What is the ZDCEnergyParam? It has a header, mentioned in the LinkDef, but has not cxx and not compiled. Commenting it at the moment.
You had lot of other stuff with ClassDefNV, mentioned in the LinkDef but not in the CMakeFiles o2_target_root_dictionary

I've opened a PR in your repo, with it the object is read correctly.

PS Use `#include "Framework/Logger.h"` instead of `#include "FairLogger.h"`

**Comment by cortesep** [comment]
Thanks for the corrections Ruben

**Comment by shahor02** [comment]
@cortesep it seems you have merged instead of rebasing to dev. 
Please rever your merge commit: `git revert 297b7fc08bd55dd4d5d141691834aa7f59202d29`
then do `git rebase dev`

**Comment by cortesep** [comment]
I cannot revert. I get this error

git revert 297b7fc08bd55dd4d5d141691834aa7f59202d29
error: Commit 297b7fc08bd55dd4d5d141691834aa7f59202d29 is a merge but no -m option was given.
fatal: revert failed

**Comment by shahor02** [comment]
@cortesep :  then do 
````
git reset --hard d7bad60056f521e05e6af555b4eeeaad282c79e5
````
i.e. reset to your last commit before the merge. And try to avoid merge in future...

**Comment by cortesep** [comment]
Tested reconstruction with:
- pythia8 simulation
- digit2raw
- raw2digit
- reconstruction
Output looks reasonable

**Comment by shahor02** [comment]
OK, thanks, will merge once you will cleanup the wrong merge, rebase to dev, forcepush and it passes the checks.

**Comment by cortesep** [comment]
It is very difficult to recover from the merge done by mistake. I'm closing this pull request and opening a new one from a fresh tree.

**Pull request closed without merging**
